### ISSUE TITLE
Training CycleGAN with 2D data 

### ISSUE THREAD

**Issue by phamnam95** [junyanz/pytorch-CycleGAN-and-pix2pix#304]
Hello. I am training CycleGAN with 2D data but the range is not from 0 to 255 like in image data. I saw in the code, the data is normalized by (data-127.5)/127.5 to be in between -1 and 1. The output from generator is activated by tanh function to be in between -1 and 1. Then the result is converted back to RGB by (output*127.5)+127.5. I am wondering if I train with 2D data in a random range, how can I train and output result. Thanks

**Comment by SsnL** [comment]
What's your "random range"

**Comment by phamnam95** [comment]
I am using seismic images. After normalizing the data by using (data-mean(data))/std(data), the range is from -10 to 10.

**Comment by SsnL** [comment]
Range is bounded it seems? Then you can just similarly normalize to [-1, +1] range, without using dataset mean and std.

**Comment by phamnam95** [comment]
Is this neccesary to put tanh as activation function in last layer? Can I use relu, lrelu or not using activation function at all? I tried normalizing the input to [-1,1] but the result is not as good as normalizing using mean, std, and not using tanh as activation function.

**Comment by SsnL** [comment]
tanh makes the result bounded. relu can potentially give you unbounded ones.

**Comment by John1231983** [comment]
@phamnam95 : So do you use tanh in last layer? What kind of normalization do you prefer? I want to use ` (data-mean(data))/std(data)` but it seem that pytorch has no layer to normalize zero mean and unit variance

**Comment by skaudrey** [comment]
Here are my data: 2D, not image, and each feature has different range.
If I did not rescale my data, then G_loss is usually at 1000 or more. Aftre rescaled my data into [-1,1], G_loss went down to 0.8, and D_loss is mostly at 0.5. Here are my questions:

1. It seems the bias between generated data and ground truth is sensbile of the range. However, not like images which I know exactly the range is [0,255], the range of my data are done by sampling approximation. So, should I concentrate on getting a better approximation of range?

2. Is it a good result with D_loss mostly at 0.5?

3. Can you give me more advice for these kind of data?

Thanks anyway.
@SsnL

**Comment by SsnL** [comment]
Generators in this repo is designed for data in range [-1, 1], as you can
see from the tanh at the end. So data with larger range will definitely
fail. Moreover, it is generally a good idea to normalize data in DL, as it
empirically stables training.

In GAN training, D loss is often not a very good metric for quality. You
should look at the results and see how they are, or evaluate them with
domain specific metrics if you have any.

**Comment by liuhh02** [comment]
> Generators in this repo is designed for data in range [-1, 1], as you can see from the tanh at the end. So data with larger range will definitely fail. Moreover, it is generally a good idea to normalize data in DL, as it empirically stables training.

@SsnL How would you suggest normalizing the data to the range [-1, 1], and "de-normalize" the data back to the original values during testing? I'm currently doing normalized_data= (data - data_avg) / data_avg where data_avg = (data.min() + data.max()) / 2 to normalize the values to [-1, 1]. But because my datatype doesn't have a fixed minimum and maximum value, I just take the maximum and minimum of the training dataset to determine the average. Is this correct? Or should I be normalizing the data based on each individual image's minimum and maximum values? Also, how can I "de-normalize" the values back to the original?
